### ISSUE TITLE
[R] More compact Roxygen

### ISSUE THREAD

**Issue by mayer79** [dmlc/xgboost#9875]
Roxygen is now closer to Markdown. E.g., instead of the verbose

```
 \itemize{
   \item \code{Features}
   ...
}
```

one can write

```

- `Features`
- ...
```

This makes the docstrings easier to read. 

I can work on this, if nobody has any objections.

- [x] callbacks.R
- [x] utils.R
- [ ] xgb.Booster.R
- [ ] xgb.config.R
- [ ] xgb.create.features.R  
- [ ] xgb.cv.R
- [ ] xgb.DMatrix.R         
- [ ] xgb.DMatrix.save.R    
- [ ] xgb.dump.R    
- [x] xgb.ggplot.R  
- [x] xgb.importance.R
- [x] xgb.load.R
- [x] xgb.load.raw.R
- [x] xgb.model.dt.tree.R
- [x] xgb.plot.deepness.R
- [x] xgb.plot.importance.R
- [x] xgb.plot.multi.trees.R 
- [x] xgb.plot.shap.R     
- [x] xgb.plot.tree.R
- [ ] xgb.save.R
- [x] xgb.save.raw.R
- [ ] ~xgb.serialize.R~
- [ ] xgb.train.R
- [ ] ~xgb.unserialize.R~
- [ ] xgboost.R

**Comment by trivialfis** [comment]
> I can work on this, if nobody has any objections.

Thank you for volunteering! Feel free to open a PR.

**Comment by trivialfis** [comment]
@mayer79 Try to use this in the XGB repo, it's using roxygen:
```
python ./tests/ci_build/test_r_package.py --task=doc
```

Yes, it does build the package.

**Comment by david-cortes** [comment]
> @trivialfis : Beginners question: How do you generate the rd help files in a convenient way?
> 
> For other packages, I usually run `devtools::document()` or `devtools::document("R-package")`.. But that starts building the library and stops with
> 
> "fatal error: common.h: No such file or directory #include <dmlc/common.h"

I think you might have cloned the repository without the submodules. You need to use `clone --recursive`.

Maybe you could try something like this:
```shell
git submodule update --init --recursive
```

**Comment by jameslamb** [comment]
> `python ./tests/ci_build/test_r_package.py --task=doc`

You can avoid needing to rebuild + reinstall every time by passing `load = 'installed'`, like this:

```shell
# build and install the library for the first time
R CMD INSTALL --with-keep.source ./R-package

# re-generate the docs
Rscript -e "roxygen2::roxygenize('R-package/', load = 'installed')"

# (go manually change some roxygen comments)

# re-generate the docs again (no reinstall)
Rscript -e "roxygen2::roxygenize('R-package/', load = 'installed')"
```

*(all commands run from the root of the repo)*

For iterating during local development, you'll probably find that quicker than running `devtools::document()` or `roxygen2::roxygenize()` in ways that lead to reinstallation.

**Comment by mayer79** [comment]
Both hints did the trick, thanks a lot!

**Comment by trivialfis** [comment]
> Do we have an up-to-date docu on how to build the DLL?


Can you build it with:

``` sh
cd R-package
R CMD INSTALL .
```

**Comment by mayer79** [comment]
Thanks for the support, and sorry for the troubles. This gives the same:

```
D:\xgboost\R-package>R CMD INSTALL .
* installing to library 'C:/Users/Michael/AppData/Local/R/win-library/4.3'
* installing *source* package 'xgboost' ...
** using staged installation
** libs
Makevars.win:19: -DXGBOOST_STRICT_R_MODE=1
Makevars.win:19: -DDMLC_LOG_BEFORE_THROW=0
Makevars.win:19: -DDMLC_ENABLE_STD_THREAD=0
Makevars.win:19: -DDMLC_DISABLE_STDIN=1
Makevars.win:19: -DDMLC_LOG_CUSTOMIZE=1
Makevars.win:19: -DDMLC_CXX11_THREAD_LOCAL=0
using C compiler: 'gcc.exe (x86_64-posix-seh, Built by strawberryperl.com project) 8.3.0'
using C++ compiler: 'g++.exe (x86_64-posix-seh, Built by strawberryperl.com project) 8.3.0'
using C++17
Makevars.win:19: -DXGBOOST_STRICT_R_MODE=1
Makevars.win:19: -DDMLC_LOG_BEFORE_THROW=0
Makevars.win:19: -DDMLC_ENABLE_STD_THREAD=0
Makevars.win:19: -DDMLC_DISABLE_STDIN=1
Makevars.win:19: -DDMLC_LOG_CUSTOMIZE=1
g++  -std=gnu++17 -I"C:/PROGRA~1/R/R-4.3.0/include" -DNDEBUG -I../..//include -I../..//dmlc-core/include -I../..//rabit/include -I../../ -DXGBOOST_STRICT_R_MODE=1 -DDMLC_LOG_BEFORE_THROW=0 -DDMLC_ENABLE_STD_THREAD=0 -DDMLC_DISABLE_STDIN=1 -DDMLC_LOG_CUSTOMIZE=1    -I"c:/rtools43/x86_64-w64-mingw32.static.posix/include"  -fopenmp -DDMLC_CMAKE_LITTLE_ENDIAN=1 -pthread    -O2 -Wall  -mfpmath=sse -msse2 -mstackrealign  -c ../..//src/collective/allgather.cc -o ../..//src/collective/allgather.o
In file included from ../..//src/collective/loop.h:17,
                 from ../..//src/collective/comm.h:14,
                 from ../..//src/collective/allgather.h:13,
                 from ../..//src/collective/allgather.cc:4:
../..//include/xgboost/collective/socket.h: In function 'xgboost::collective::Result xgboost::collective::INetNToP(const H&, std::__cxx11::string*)':
../..//include/xgboost/collective/socket.h:715:7: error: there are no arguments to 'inet_ntop' that depend on a template parameter, so a declaration of 'inet_ntop' must be available [-fpermissive]
       inet_ntop(AF_INET, addr, str, INET_ADDRSTRLEN);
       ^~~~~~~~~
../..//include/xgboost/collective/socket.h:715:7: note: (if you use '-fpermissive', G++ will accept your code, but allowing the use of an undeclared name is deprecated)
../..//include/xgboost/collective/socket.h:722:7: error: there are no arguments to 'inet_ntop' that depend on a template parameter, so a declaration of 'inet_ntop' must be available [-fpermissive]
       inet_ntop(AF_INET6, addr, str, INET6_ADDRSTRLEN);
       ^~~~~~~~~
make: *** [C:/PROGRA~1/R/R-4.3.0/etc/x64/Makeconf:270: ../..//src/collective/allgather.o] Error 1
ERROR: compilation failed for package 'xgboost'
* removing 'C:/Users/Michael/AppData/Local/R/win-library/4.3/xgboost'
* restoring previous 'C:/Users/Michael/AppData/Local/R/win-library/4.3/xgboost'
```

Maybe try with a different compiler?

**Comment by trivialfis** [comment]
That's interesting, we run the command on CI https://github.com/dmlc/xgboost/blob/ae32936ba242e3fe083f79e8592b337220c1cb73/tests/ci_build/test_r_package.py#L221

**Comment by trivialfis** [comment]
The R distribution I use on Windows is downloaded from Rtools official site. https://cran.r-project.org/bin/windows/Rtools/rtools43/rtools.html

**Comment by mayer79** [comment]
You got me... I was still on rtools40 😇

**Comment by trivialfis** [comment]
That's even more interesting, you have a line in the log:

> g++  -std=gnu++17 -I"C:/PROGRA~1/R/R-4.3.0/include" -DNDEBUG -I../..//include -I../..//dmlc-core/include -I../..//rabit/include -I../../ -DXGBOOST_STRICT_R_MODE=1 -DDMLC_LOG_BEFORE_THROW=0 -DDMLC_ENABLE_STD_THREAD=0 -DDMLC_DISABLE_STDIN=1 -DDMLC_LOG_CUSTOMIZE=1    -I"c:/rtools43/x86_64-w64-mingw32.static.posix/include"  -fopenmp -DDMLC_CMAKE_LITTLE_ENDIAN=1 -pthread    -O2 -Wall  -mfpmath=sse -msse2 -mstackrealign  -c ../..//src/collective/allgather.cc -o ../..//src/collective/allgather.o

**Comment by mayer79** [comment]
At least the path variable was pointing to C:/rtools40... let's see what I get with clean environment variables.

**Comment by trivialfis** [comment]
> A different error message pops up: Is g++ version too high?

No, I don't think the compiler version can be "too high" for C++ code. Judging from:

> C:\rtools43\x86_64-w64-mingw32.static.posix\bin/ld.exe: ../..//src/data/file_iterator.o:file_iterator.cc:(.rdata$_ZTVSt12codecvt_utf8IwLm1114111ELSt12codecvt_mode0EE[_ZTVSt12codecvt_utf8IwLm1114111ELSt12codecvt_mode0EE]+0x20): undefined reference to `std::__codecvt_utf8_base<wchar_t>::do_out(int&, wchar_t const*, wchar_t const*, wchar_t const*&, char*, char*, char*&) const'

I think you have a mix of R environments and the compiler is using the wrong standard library.

**Comment by mayer79** [comment]
@trivialfis Thanks your your patient assistance! A simple `--preclean` was now sufficient to build the library :-). 

```
R CMD INSTALL --preclean --with-keep.source ./R-package
```

**Comment by trivialfis** [comment]
Excellent! That's probably caused by the binaries compiled before you upgrade your toolchain to 4.3. If I guess right, the second run after the first successful compilation should not require the preclean option.

**Comment by mayer79** [comment]
Implemented in #9906, #9919, #10674, and #10733
